[Feature] 회원가입 이메일 인증 기능 구현

zoooooz2 opened the pull request:
## 📌 관련 이슈
- close #97 

## 📝 변경 사항
### AS-IS
-  회원가입할 때, 이메일 인증, 검사 기능 부재
- 이메일 에러에 맞는 에러코드 부재
- 인증번호 검사하는 기능 부재!
- 이메일 인증번호 전송 & 인증 테스트 코드 부재
- 최종 회원가입 시에 모든 값을 검증하는 코드 부재

### TO-BE
- 사용자가 이메일 인증을 누름 -> 스프링 서버 요청 -> 이메일 서버 (SMTP서버)가 인증번호 를 보내줌 -> 인증번호 DB에 저장
-  인증번호는 랜덤으로 생성, 만료 시간 3분(`MailService`, `.yml`)
- 이메일 형식 체크, 예외처리(`MailService`)
- 에러코드 6개 추가(`ErrorCode`)
- 인증번호 비교를 위해 엔티티 추가(EmailAuthCode)
- 인증번호 비교&검사(DB에 있는지-이메일, 인증번호/만료시간 확인)
- 인증번호 받고 만료 확인 후, 예외 던지는 테스트 코드 작성
- 최종 회원가입 로직에 모든 값(이름, 이메일, 인증번호)을 검증하는 코드 추가 - 비밀번호는 DTO에서
- 최종 회원가입 요청 DTO에 인증번호(authCode) 추가

## 🔍 테스트
- [X] 테스트 코드 작성
- [x] API 테스트 - 성공, 실패(`E002`, `E004`) 확인
- [x] 로컬 테스트 - 오류x

## 📸 스크린샷
<img width="538" height="513" alt="image" src="https://github.com/user-attachments/assets/9f00dba2-f8a2-417a-901d-e69456a94353" />
<img width="476" height="532" alt="image" src="https://github.com/user-attachments/assets/393061ac-419a-4f34-94f7-c40a6cf5d5c2" />
<img width="492" height="538" alt="image" src="https://github.com/user-attachments/assets/0427b8e5-3973-4ec8-b56c-b0d99fb327c1" />
<img width="597" height="617" alt="스크린샷 2025-10-22 190456" src="https://github.com/user-attachments/assets/fc767a17-a494-481d-a1e0-016ea925c083" />

## ✅ 체크리스트
- [x] main/develop 브랜치가 아닌 feature 브랜치에서 작성했나요?
- [x] 코딩 컨벤션을 준수했나요?
- [x] 불필요한 코드는 제거했나요?
- [x] 주석은 충분히 작성했나요?
- [x] 테스트는 완료했나요?

## 🙋‍♂️ 리뷰어에게
- 환경변수 값은 줌 채팅에 보내겠습니다~
- Gmail(javaMailSender) 사용했습니다.
- 응답 값 - 성공: true 실패:에러코드
- postman으로 통합테스트 과정 사진 https://www.notion.so/api-2959d0051b99803fa349c144c456835e?source=copy_link
- 커밋을 못 나눠서,, 커밋의 하위(상세) 설명을 보시면 될 것 같습니다.
<img width="660" height="327" alt="image" src="https://github.com/user-attachments/assets/b5827e94-21f4-4463-a314-34efb8957bd9" />

***코드가 난리법석인데 급해서 구현만 되게끔 만든 점 양해 부탁드립니다,,,,***

zoooooz2 committed:
feat: email 의존성 추가

zoooooz2 committed:
docs: yml에 email 속성 추가

zoooooz2 committed:
chore: yml의 email 속성값을 환경변수로 수정

zoooooz2 committed:
feat: 이메일 인증 기능(인증번호 전송) 구현

- 컨트롤러 추가
- global > email > MailConfig, MailService 추가
- 이메일 관련 ErrorCode 추가
- 이메일 예외처리, 형식 체크

zoooooz2 committed:
Merge branch 'develop' into feature/email

dpwl0974 commented on backend/src/main/java/com/backend/petplace/domain/email/service/EmailAuthCodeService.java:
```diff
@@ -0,0 +1,86 @@
+    try {
+      message.setFrom(senderEmail);
+      message.setRecipients(MimeMessage.RecipientType.TO, mail);
+      message.setSubject(TITLE);
+      String body = "";
+      body += "<h3>요청하신 인증 번호입니다.</h3>";
+      body += "<h1>" + authCode + "</h1>";
+      body += "<h3>감사합니다.</h3>";
+      message.setText(body, "UTF-8", "html");
```
이 부분은 나중에 리팩토링 하실 때,  코드 가독성 및 확장성을 고려하여  따로 html로 빼면 좋을 것 같습니다 !
확장성의 예를 들면, 비밀번호 찾기 등에도 이메일 인증이 쓰이니 그럴 땐 용도별로 나눠서 이메일을 보내면 어떨까 싶습니다 :)

zoooooz2 replied:
저도 해야되나 생각했는데 확실히 하는 게 좋겠군요!! 좋은 피드백 감사합니다! 나중에 이 부분도 수정해보겠습니다~!

dpwl0974 commented on backend/src/main/resources/application.yml:
```diff
@@ -25,6 +25,22 @@ spring:
+          connection timeout: 5000
+          timeout: 5000
+          write timeout: 5000
+      auth-code-expiration0millis: 300000  # 30 * 60 * 1000(밀리초) == 30분
```
이메일 인증 유효 시간이 좀 길게 느껴지는 것 같습니다..!

zoooooz2 replied:
혹시 몇 분으로 하는 게 좋다고 생각하시나요?? 5분 이하??

dpwl0974 replied:
네! 한 3~5분 사이면 괜찮지 않을까요?

zoooooz2 replied:
알겠습니다! 좋은 것 같습니다 피드백 감사합니다~!

zoooooz2 committed:
feat: 이메일 만료 3분으로 단축

zoooooz2 committed:
feat: 이메일 인증번호 확인 기능 추가

- global > email > MailService 파일 위치 변경
- 인증번호를 DB에 저장하고 확인하기 위해 엔티티 추가(EmailAuthCode)

- 서비스 로직
- 중복을 막기 위해 인증번호 생성할 때, 개수 +1, 소문자도 추가
- @Email이 검사해준다고 해서 ValidateMail? 메서드 삭제
- sendMail하고나서 인증번호, 이메일 DB에 저장
- 인증번호 체크하는 메서드 추가(DB에 있는지, 만료시간 체크) + 인증 성공하면 데이터 삭제

- 컨트롤러
- 인증번호 체크하는 컨트롤러 추가(Patch) - Body로 값 보냄
- 요청 DTO 생성 - email, authCode

zoooooz2 committed:
test: 인증번호 받고 만료 예외 던지는 테스트 코드 작성 + .yml 시간 수정

zoooooz2 committed:
style: jwt 테스트(단위 -> 통합) 주석 수정

dpwl0974 reviewed (dismissed):
고생하셨습니다 :)

hodakrer reviewed (dismissed):
서기님 고생 많으셨어요!

DEV-Cheeze reviewed (dismissed):
수고하셨습니다!
혹시 원리가 요청값 오면 인증코드 db에 올려놨다가 사용자 입력값 검증할때 db와 같이 검증해서 통과시키는건가요 ?

zoooooz2 committed:
Merge branch 'develop' into feature/email

zoooooz2 committed:
feat: 이름, 이메일 중복 확인 요청을 위해 api 주소 분리, 응답 DTO 추가

- 각각 버튼을 따로 프론트에서 만들 예정이라서 api 주소 추가해주었습니다.

zoooooz2 committed:
style: 코드 줄 띄어쓰기 수정

zoooooz2 committed:
feat: 메일 인증번호 전송, 검사하는 로직 메서드에 @Transactional 추가, 반환값 응답 DTO(BoolResultResponse)로 수정

zoooooz2 committed:
refactor: 회원 추가 기능을 정적 팩토리 메서드로 변경

zoooooz2 committed:
feat: 회원가입 최종 완료 요청 시, 모든 값 검증하는 로직 추가

email
- 인증번호 인증 요청DTO에 메시지 추가
- EmailAuthCode 엔티티에 인증 확인 값 추가(verified)
- 인증번호 체크할 때 성공하면 삭제하지 않고, verified값을 true로 변경

UserController - /signup
- 검증을 위해, 회원가입 요청DTO에 인증번호 필드 추가

UserService - signup()
- 이름, 이메일 중복 검사 메서드 추가
- (참고: 이름, 비밀번호, 이메일의 유효성 검사는 요청DTI에서 처리)
- 이메일 인증 체크 메서드 추가(이메일, 인증번호를 같이 비교하여 존재
  확인, 인증성공: DB에서 인증번호 삭제 / 실패: 에러처리(AC003)

dbghwns123 commented on backend/src/main/java/com/backend/petplace/domain/email/entity/EmailAuthCode.java:
```diff
@@ -0,0 +1,58 @@
+        .email(email)
+        .authCode(authCode)
+        .expiredAt(LocalDateTime.now().plusMinutes(authCodeExpirationTime))
+        .verified(false)
```
여기서 verified로 검증하는 방식 좋은 것 같습니다!

zoooooz2 replied:
네!! 피드백 감사합니다~!!!!

dbghwns123 reviewed (approved):
고생 많으셨습니다~~

dpwl0974 reviewed (approved):
고생하셨습니다 :)

hodakrer reviewed (approved):
고생많으셨습니다.

DEV-Cheeze reviewed (approved):
정말 고생 많이 하셨습니다...!
